[JIT] Adds fp16 support to the jit

mruberry opened the pull request:
This PR adds support for fp16 tensors to the jit, and it adds a test to ensure they're working properly.

Half tensors are treated just like in reductions. Half inputs are immediately upconverted to float, mathematical operations are performed only on floats, and floats are downconverted to half when writing to half outputs.

To add support for half tensors I extended the (renamed) "SupportedTypes" to include halfs, created a snippet with the appropriate half conversions defined and included it as a string literal (this snippet courtesy of @ngimel), then check for half types and write the appropriate access line.

For testing I added a new test and (very slightly) refactored three existing tests to reuse the functions they defined. Rerunning each test in test_jit.py for every datatype seems excessive, especially since many of the functions deal with graph structure. Further, half types are different from other data types since we cannot simply run a half tensor through an unfused version of the same operations and expect the same final result as the fused version. Instead, a float tensor that's representable in half must be passed through the unfused network while an equivalent tensor in half is passed through the fused model, and the results of the unfused network cast to half for comparison. This process mimics the conversions done by the fused kernel and is tricky to merge with existing code paths without greatly complicating them.

mruberry committed:
adds fp16 support to the jit

mruberry committed:
improves formatting

mruberry committed:
improves formatting

mruberry committed:
added an explanatory comment

mruberry committed:
fixes Python2 flake8

mruberry commented:
The failures on pytorch-linux-trusty-py3.6-gcc7.2 are interesting as they appear isolated to that system but are in the JIT. In particular, the JIT is complaining that it cannot compile some CPU kernels (and suggesting some assembly arguments are wrong!). 

The same tests run correctly on pytorch-linux-trusty-py3.6-gcc5.4, suggesting this issue is related to the gcc version. 

I did not expect these tests or CPU kernels to be affected by these changes. I will look at setting up an environment to review them more carefully but if someone knows a particular gcc7.2 gotcha I missed please let me know.

ezyang commented:
If this were an older version of GCC, I would have just advised you to X out the test on that test configuration. But it's a newer version so that implies we might have to talk to upstream. Can you extract out the C program that triggers the error?

mruberry commented:
@ezyang Thanks for your thinking; give me a moment to investigate. This is (almost) definitely regressed as a result of my change (since other CIs are passing fine). I'm reviewing the generated kernels now.

mruberry committed:
updates c code

mruberry commented:
This PR is now ready for review.

I confess I do not know exactly what 7.2 demon I upset, but the following updates resolved the issue:

(1) I moved the half header template hook into the CUDA ifdef for safety (I verified it is not appearing in c generated code, but it was emitting an extra blank line)
(2) I removed the iostream include from the C++ kernels and added cstdint, this is a bug fix unrelated to the original PR since those kernels were (apparently) relying on iostream including cstdint
(3) I added the -lm linking directive to the gcc invocation since the C++ kernels include math.h

ezyang commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -121,6 +122,44 @@ void ${kernelName}(IndexType totalElements, void ** args) {
+    /* Definitions of intrinsics */
+    __device__ __half __float2half(const float f) {
+      __half val;
+      asm("{  cvt.rn.f16.f32 %0, %1;}\n" : "=h"(__HALF_TO_US(val)) : "f"(f));
```
Oh man! Is this actually stable, as in, CUDA 10 isn't going to change this out from under us?

mruberry replied:
We'll make sure this change works with CUDA 10.

ezyang commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -159,6 +198,10 @@ std::string valueName(Value * n) {
 const char * scalarTypeName(at::ScalarType type) {
+  if (type == at::ScalarType::Half) {
+    return "half";
+  }
```
Maybe would be even clearer if we switch the macro below from `AT_FORALL_SCALAR_TYPES` to `AT_FORALL_SCALAR_TYPES_EXCEPT_HALF`.

mruberry replied:
That would be clearer; give me a moment and I'll fix that up.

ezyang reviewed (approved):
Looks legit.

apaszke commented on test/test_jit.py:
```diff
@@ -416,27 +419,65 @@ def f(x, y):
+            outputs = fn(*local_inputs)
+            fusion_outputs = fusion(*local_fusion_inputs)
+            outputs_half = [t.half() for t in outputs]
+            self.assertEqual(outputs_half, fusion_outputs)
```
This assert is very confusing for me. If I understand correctly, you simply enabled the half parts of the graph as eligible for fusion, but in here you effectively run the whole graph if `float` (even using `fusion`). I can't see any `half` casts except the `outputs_half` one. Is this really checking the new codepath?

mruberry replied:
I believe you are missing the dtype=torch.half when x and y are created.

The fusion_inputs tuple is composed of half tensors and they create the local_fusion_inputs which are put through the torch.jit.trace call. The inputs tuple is float tensors that are exactly representable in half. 

As I mention above, it would be risky to run halfs through an unfused network because there could be loss of precision within and across operations. That's why I'm doing the work here to compare floats representable as halfs through an unfused network vs the equivalent halfs through the fused network.

apaszke replied:
Ohhh ok. Yes, I fully understand why are we running the unfused thing in float, I just searched for more `half()` calls, but missed the `dtype` part which was too far up for me to see from here. Never mind the comment. LGTM!

mruberry committed:
all except halfs

mruberry commented:
The failure on pytorch-linux-trusty-py3.6-gcc5.4 seems unrelated to the macro change. It looks like ONNXBot updated ir.h to use std::sqrt and that's complaining, possibly because it's including math.h (which may not define names in std).
